### PR TITLE
Controllers : Add a new kind onf controller to expose ReadOnly state from the application

### PR DESCRIPTION
**Pull request by samcake**
- A new "StateController" that exposes read only endpoints to the userInputMapper that are just REadOnly lambdas.

- in APplication we added the ApplicationStateDevice that is a StateCOntroller and that exposes for now a single new input called "InHMD" that can be queried to know if the application is i"INHMD" or not

### CODE REVIEW

File: examples/controllers/handPosesDebug.js [hunk 1]
```diff
@@ -84,9 +85,46 @@ function updateHand(handNum, deltaTime) {
 	}
 }
 
+function updateHydra(handNum, deltaTime) {
```
Comment (ZappoMan) [inline examples/controllers/handPosesDebug.js]: formatting is wrong in this file... you have TABS

File: interface/resources/controllers/standard.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
     "channels": [
         { "from": "Standard.LY", "to": "Actions.TranslateZ" },
         { "from": "Standard.LX", "to": "Actions.TranslateX" },
-        { "from": "Standard.RX", "to": "Actions.Yaw" },
+         { "from": "Standard.RX", "to": "Actions.Yaw" },
```
Comment (ZappoMan) [inline interface/resources/controllers/standard.json]: extra space?

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -633,9 +635,20 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer) :
...
+        return (float) qApp->getAvatarUpdater()->isHMDMode();
+    });
+    _applicationStateDevice->addInputVariant("InHMD", InHMDLambda);
+
+    userInputMapper->registerDevice(_applicationStateDevice);
+    
     // Setup the keyboardMouseDevice and the user input mapper with the default bindings
     userInputMapper->registerDevice(_keyboardMouseDevice);
 
+    //userInputMapper->getApplicationDevice()->(float)qApp->getAvatarUpdater()->isHMDMode()
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: dead code?

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -2738,6 +2755,11 @@ void Application::update(float deltaTime) {
         }
         myAvatar->setDriveKeys(ZOOM, userInputMapper->getActionState(controller::Action::TRANSLATE_CAMERA_Z));
     }
+
+    float lhc = userInputMapper->getActionState(controller::Action::LEFT_HAND_CLICK);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: dead code?

File: libraries/controllers/src/controllers/Actions.cpp [hunk 5]
```diff
@@ -75,6 +75,8 @@ namespace controller {
                 makeAxisPair(Action::BOOM_IN, "BoomIn"),
                 makeAxisPair(Action::BOOM_OUT, "BoomOut"),
 
+                makeButtonPair(Action::IN_HMD, "InHMD"),
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/Actions.cpp]: dead code? I thought we agreed this wasn't an action?

File: libraries/controllers/src/controllers/Actions.h [hunk 6]
```diff
@@ -78,6 +78,8 @@ enum class Action {
     BOOM_IN,
     BOOM_OUT,
 
+    IN_HMD, // THis is a read only action ?
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/Actions.h]: dead code? I thought we agreed this wasn't an action?

File: libraries/controllers/src/controllers/StateController.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    StateController(QString name);
+    virtual ~StateController();
+
+    using ReadLambda = std::function<float()>;
+    using NamedReadLambda = QPair<QString, ReadLambda>;
+
+    void addInputVariant(QString name, ReadLambda& lambda);
+
+    QVector<NamedReadLambda> namedReadLambdas;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/StateController.h]: why are these public? shouldn't they be private and with _

File: libraries/controllers/src/controllers/StateController.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    proxy->_name = _name;
+    proxy->getButton = [this] (const Input& input, int timestamp) -> bool { return getButton(input.getChannel()); };
+    proxy->getAxis = [this] (const Input& input, int timestamp) -> float { return getAxis(input.getChannel()); };
+    proxy->getAvailabeInputs = [this] () -> QVector<Input::NamedPair> {
+    
+        
+        QVector<Input::NamedPair> availableInputs;
+        
+        int i = 0;
+        for (auto pair : namedReadLambdas) {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/StateController.cpp]: shouldn't this be `const auto& pair`? you don't really need a value for the pair, when a const& will do

**Comment by hifi-gustavo** [comment]
Build console output for commit: fc15c7cd98f147f0b7c410904ff460bf26f6a2aa available at: https://gist.github.com/463eb117efd226e790d6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.dmg

File: libraries/controllers/src/controllers/Actions.cpp (hunk 5 above)
Comment (samcake) [inline libraries/controllers/src/controllers/Actions.cpp]: good point, that was my first take on this that remained, i m removing it

File: libraries/controllers/src/controllers/StateController.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    StateController(QString name);
+    virtual ~StateController();
+
+    using ReadLambda = std::function<float()>;
+    using NamedReadLambda = QPair<QString, ReadLambda>;
+
+    void addInputVariant(QString name, ReadLambda& lambda);
+
+    QVector<NamedReadLambda> _namedReadLambdas;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/StateController.h]: should be private?

**Comment by hifi-gustavo** [comment]
Build console output for commit: b37a6f689a2e2478409cddff5885b4bf2f9b3ca3 available at: https://gist.github.com/463eb117efd226e790d6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bbc6d9f5e7709427d1ab37b9ac4b9124dcc03a34 available at: https://gist.github.com/463eb117efd226e790d6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 07c13f4d660f5488cf3d0b34e3f2d85d094aad1c available at: https://gist.github.com/463eb117efd226e790d6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b90af1a1ce636f6dc66d098011b8a8ac508643ab available at: https://gist.github.com/463eb117efd226e790d6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6206.dmg

**Pull request merged**